### ISSUE TITLE
DNS resolution for broker address

### ISSUE THREAD

**Issue by ShrirAm-cse** [IBM/sarama#2474]
##### TL;DR :
How to configure sarama producer to perform an equivalent action of 
`client.dns.lookup = "use_all_dns_ips"`
[reference link](https://cwiki.apache.org/confluence/display/KAFKA/KIP-602%3A+Change+default+value+for+client.dns.lookup)

##### Scenario :
Suppose I am using a producer and I provide a list of broker addresses in the form of domain name ( example - abc.xyz.co )  to  the constructor.
Lets say, I have provided two domain names DN1 and DN2 as the broker addresses to NewSyncProducer function and each domain name resolves to multiple IPs ( example - DN1 resolves to ipA,ipB and DN2 resolves to ipC,ipD,ipE ).
Is there a way to make sarama to try to connect every resolved ip of a domain name before moving on to the next domain name for choosing it as a producer?

##### Problem :
The Sarama producer in Golang caches a single IP address initially when resolving the broker's hostname. This cached IP address is retrieved from the cachedLeader method in the client.go file.

The issue arises when the second resolution attempt occurs and the previously cached IP address is no longer valid or becomes unreachable. As a result, the program terminates due to the inability to establish a connection.

**Comment by dnwe** [comment]
@ShrirAm-cse `use_all_dns_ips` should essentially be the only behaviour that Sarama has. 

When you provide a list of domain names to the NewSyncProducer, those are the bootstrap addresses which are used to send the initial ApiVersionsRequest and request(s) for the cluster Metadata (i.e., the available brokers in the cluster and where topicpartitions are hosted). These connections _may_ get re-used by the producer, but it depends which broker(s) are the active leaders for the topic partitions that your producer is sending records to. This is the only stage at which a kafka client will use multiple domain names whilst trying to make its initial connection.

The `cachedLeader` func you mention returns a Broker instance rather than an IP address. Each broker instance will have an associated domain name as returned by the `advertised.listeners` field in your server-side configuration and connectivity to it will be (re-)established using the start net.Dialer which will attempt each resolved IP address in turn. For any given port a broker will only have one advertised hostname
